server_test.go

```diff
@@ -21,8 +21,195 @@ func TestNewServer(t *testing.T) {
+			quitTimeout:   make(chan interface{}),
+		}
+		listener.On("Close").Once().Return(nil)
+		waitGroup.On("Wait").Once().Return(nil)
+		logger.On("infoActivity", serverStopMsg).Once().Return(nil)
+
+		assert.NoError(t, server.Stop())
+		assert.False(t, server.isStarted())
+		_, isChannelOpened := <-server.quit
+		assert.False(t, isChannelOpened)
+	})
+
+	t.Run("when server active stops current server, force shutdown case", func(t *testing.T) {
+		logger, listener, waitGroup, quitChannel := new(loggerMock), new(listenerMock), new(waitGroupMock), make(chan interface{})
+		server := &Server{
+			configuration: createConfiguration(),
+			logger:        logger,
+			listener:      listener,
+			wg:            waitGroup,
+			quit:          quitChannel,
+			started:       true,
+		}
+		listener.On("Close").Once().Return(nil)
+		waitGroup.On("Wait").Once().Return(nil)
+		logger.On("infoActivity", serverForceStopMsg).Once().Return(nil)
+
+		assert.NoError(t, server.Stop())
+		assert.False(t, server.isStarted())
+		_, isChannelOpened := <-server.quit
+		assert.False(t, isChannelOpened)
+	})
+
+	t.Run("when server is inactive doesn't stop current server", func(t *testing.T) {
+		assert.EqualError(t, new(Server).Stop(), serverStopErrorMsg)
+	})
+}
+
+func TestServerMessages(t *testing.T) {
+	configuration := createConfiguration()
+
+	t.Run("when there are no messages on the server", func(t *testing.T) {
+		server := newServer(configuration)
+
+		assert.Empty(t, server.Messages())
+	})
+
+	t.Run("when there are messages on the server", func(t *testing.T) {
+		server := newServer(configuration)
+		server.newMessage()
+
+		assert.NotEmpty(t, server.Messages())
+	})
+
+	t.Run("confirm message data of internal server.messages.items and exported Messages() slice are identical for a given moment", func(t *testing.T) {
+		server := newServer(configuration)
+
+		assert.Empty(t, server.messages.items)
+		assert.Empty(t, server.Messages())
+		assert.NotSame(t, server.messages.items, server.Messages())
+
+		message := server.newMessage()
+
+		assert.Equal(t, []*Message{message}, server.messages.items)
+		assert.Equal(t, []Message{*message}, server.Messages())
+		assert.NotSame(t, server.messages.items, server.Messages())
+	})
+}
+
+func TestServerPortNumber(t *testing.T) {
+	t.Run("returns server port number", func(t *testing.T) {
+		portNumber := 2525
+		server := &Server{portNumber: portNumber}
+
+		assert.Equal(t, portNumber, server.PortNumber())
+	})
+}
+
+func TestServerIsStarted(t *testing.T) {
+	t.Run("returns current server started-flag status", func(t *testing.T) {
+		server := &Server{started: true}
+
+		assert.True(t, server.isStarted())
+	})
+}
+
+func TestServerSetListener(t *testing.T) {
+	t.Run("sets server listener", func(t *testing.T) {
+		server := new(Server)
+		listener, _ := net.Listen("tcp", "localhost:2526")
+		server.setListener(listener)
+
+		assert.Equal(t, listener, server.listener) // TODO: do we need thread-safe getter for server.listener ?
```

Comment:
@benjamin-rood What do you think about this poimt?

Reply:
I did consider this, but I don't think it's necessary, because we aren't ever reading from it (like this) outside of the unit tests. You *must* do this read of `server.listener` in this unit test to confirm that the `setListener` method is working, but you are never doing it anywhere else and the listener is never accessible by users of the library/package.

I think, as a rule of thumb, so long as a unit test is the only place you ever do it AND you are explicit about it in the test code then it's completely fine. Here you are clearly comparing the created `listener` with `server.listener`, and aren't doing some indirection however slight somewhere else—e.g. if you did `x := server.listener` 100 lines further up the code and then checked equality with `listener` and `x`—then you are making the tests a bit harder to validate/review by eye for correctness, and it opens up the possibility for the value of `x` to change in-between the initialisation and the call to `assert.Equal`.

This same reasoning is why I explicitly changed the message list test coverage to read from `server.messages.items` every single time instead of `messages := server.messages.items` and then doing checks on `messages`. It makes the tests slightly more ambiguous when we have a `Messages()` method on `server` which copies from `server.messages.items, for absolutely no benefit other than typing less characters.

I'm always trying to make things as blatant as possible because I am stupid. If I have to think about the code I might get it wrong or misunderstand it; if that happens, but it "seems to work" or "the tests pass" then I will ignore that bit of code, because I am lazy.

Reply:
It is one of my points to cover via tests not exported (private) methods/functions too. And one of points to explain implementation via tests. And one more big point to hate table tests approach 😃